core/src/main/java/io/substrait/extended/expression/ProtoExtendedExpressionConverter.java

```diff
@@ -0,0 +1,84 @@
+import io.substrait.expression.proto.ProtoExpressionConverter;
+import io.substrait.extension.*;
+import io.substrait.proto.ExpressionReference;
+import io.substrait.proto.NamedStruct;
+import io.substrait.type.ImmutableNamedStruct;
+import io.substrait.type.Type;
+import io.substrait.type.proto.ProtoTypeConverter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/** Converts from {@link io.substrait.proto.ExtendedExpression} to {@link ExtendedExpression} */
+public class ProtoExtendedExpressionConverter {
+  private final SimpleExtension.ExtensionCollection extensionCollection;
+
+  public ProtoExtendedExpressionConverter() throws IOException {
+    this(SimpleExtension.loadDefaults());
+  }
+
+  public ProtoExtendedExpressionConverter(SimpleExtension.ExtensionCollection extensionCollection) {
+    this.extensionCollection = extensionCollection;
+  }
+
+  private final ProtoTypeConverter protoTypeConverter =
+      new ProtoTypeConverter(
+          new ExtensionCollector(), ImmutableSimpleExtension.ExtensionCollection.builder().build());
+
+  public ExtendedExpression from(io.substrait.proto.ExtendedExpression extendedExpression) {
+    // fill in simple extension information through a discovery in the current proto-extended
+    // expression
+    ExtensionLookup functionLookup =
+        ImmutableExtensionLookup.builder()
+            .from(extendedExpression.getExtensionUrisList(), extendedExpression.getExtensionsList())
+            .build();
+
+    NamedStruct baseSchemaProto = extendedExpression.getBaseSchema();
+    io.substrait.type.NamedStruct namedStruct = convertNamedStrutProtoToPojo(baseSchemaProto);
+
+    ProtoExpressionConverter protoExpressionConverter =
+        new ProtoExpressionConverter(
+            functionLookup, this.extensionCollection, namedStruct.struct(), null);
+
+    List<ExtendedExpression.ExpressionReference> expressionReferences = new ArrayList<>();
+    for (ExpressionReference expressionReference : extendedExpression.getReferredExprList()) {
+      Expression expressionPojo =
+          protoExpressionConverter.from(expressionReference.getExpression());
+      expressionReferences.add(
+          ImmutableExpressionReference.builder()
+              .referredExpr(expressionPojo)
+              .addAllOutputNames(expressionReference.getOutputNamesList())
+              .build());
+    }
+
+    ImmutableExtendedExpression.Builder builder =
+        ImmutableExtendedExpression.builder()
+            .referredExpr(expressionReferences)
+            .advancedExtension(
+                Optional.ofNullable(
+                    extendedExpression.hasAdvancedExtensions()
+                        ? extendedExpression.getAdvancedExtensions()
+                        : null))
+            .baseSchema(namedStruct);
+    return builder.build();
+  }
+
+  private io.substrait.type.NamedStruct convertNamedStrutProtoToPojo(NamedStruct namedStruct) {
```

Comment:
```suggestion
  private io.substrait.type.NamedStruct convertNamedStructProtoToPojo(NamedStruct namedStruct) {
```

Reply:
Would it make sense to implement this function as `fromProto` in this file? https://github.com/substrait-io/substrait-java/blob/main/core/src/main/java/io/substrait/type/NamedStruct.java

Reply:
Changed